// @drift-labs/sdk/lib/constants/spotMarkets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpotMarkets = exports.MainnetSpotMarkets = exports.DevnetSpotMarkets = exports.WRAPPED_SOL_MINT = void 0;
const web3_js_1 = require("@solana/web3.js");
const __1 = require("../");
const numericConstants_1 = require("./numericConstants");
exports.WRAPPED_SOL_MINT = new web3_js_1.PublicKey('So11111111111111111111111111111111111111112');
exports.DevnetSpotMarkets = [
    {
        symbol: 'USDC',
        marketIndex: 0,
        oracle: new web3_js_1.PublicKey('5SSkXsEKQepHHAewytPVwdej4epN1nxgLVM84L4KXgy7'),
        oracleSource: __1.OracleSource.PYTH_STABLE_COIN,
        mint: new web3_js_1.PublicKey('8zGuJQqwhZafTah7Uc7Z4tXRnguqkn5KLFAP8oV6PHe2'),
        precision: new __1.BN(10).pow(numericConstants_1.SIX),
        precisionExp: numericConstants_1.SIX,
    },
    {
        symbol: 'SOL',
        marketIndex: 1,
        oracle: new web3_js_1.PublicKey('J83w4HKfqxwcq3BEMMkPFSppX3gqekLyLJBexebFVkix'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey(exports.WRAPPED_SOL_MINT),
        precision: numericConstants_1.LAMPORTS_PRECISION,
        precisionExp: numericConstants_1.LAMPORTS_EXP,
        serumMarket: new web3_js_1.PublicKey('8N37SsnTu8RYxtjrV9SStjkkwVhmU8aCWhLvwduAPEKW'),
        phoenixMarket: new web3_js_1.PublicKey('78ehDnHgbkFxqXZwdFxa8HK7saX58GymeX2wNGdkqYLp'),
    },
    {
        symbol: 'BTC',
        marketIndex: 2,
        oracle: new web3_js_1.PublicKey('HovQMDrbAgAYPCmHVSrezcSmkMtXSSUsLDFANExrZh2J'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey('3BZPwbcqB5kKScF3TEXxwNfx5ipV13kbRVDvfVp5c6fv'),
        precision: new __1.BN(10).pow(numericConstants_1.SIX),
        precisionExp: numericConstants_1.SIX,
        serumMarket: new web3_js_1.PublicKey('AGsmbVu3MS9u68GEYABWosQQCZwmLcBHu4pWEuBYH7Za'),
    },
];
exports.MainnetSpotMarkets = [
    {
        symbol: 'USDC',
        marketIndex: 0,
        oracle: new web3_js_1.PublicKey('Gnt27xtC473ZT2Mw5u8wZ68Z3gULkSTb5DuxJy7eJotD'),
        oracleSource: __1.OracleSource.PYTH_STABLE_COIN,
        mint: new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v'),
        precision: numericConstants_1.QUOTE_PRECISION,
        precisionExp: numericConstants_1.QUOTE_PRECISION_EXP,
    },
    {
        symbol: 'SOL',
        marketIndex: 1,
        oracle: new web3_js_1.PublicKey('H6ARHf6YXhGYeQfUzQNGk6rDNnLBQKrenN712K4AQJEG'),
        oracleSource: __1.OracleSource.PYTH,
        mint: new web3_js_1.PublicKey(exports.WRAPPED_SOL_MINT),
        precision: numericConstants_1.LAMPORTS_PRECISION,
        precisionExp: numericConstants_1.LAMPORTS_EXP,
        serumMarket: new web3_js_1.PublicKey('8BnEgHoWFysVcuFFX7QztDmzuH8r5ZFvyP3sYwn1XTh6'),
        phoenixMarket: new web3_js_1.PublicKey('4DoNfFBfF7UokCC2FQzriy7yHK6DY6NVdYpuekQ5pRgg'),
    },
];
exports.SpotMarkets = {
    devnet: exports.DevnetSpotMarkets,
    'mainnet-beta': exports.MainnetSpotMarkets,
};
