// @drift-labs/sdk/lib/phoenix/phoenixSubscriber.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhoenixSubscriber = void 0;
const web3_js_1 = require("@solana/web3.js");
const phoenix_sdk_1 = require("@ellipsis-labs/phoenix-sdk");
const numericConstants_1 = require("../constants/numericConstants");
const anchor_1 = require("@coral-xyz/anchor");
class PhoenixSubscriber {
    constructor(config) {
        this.connection = config.connection;
        this.programId = config.programId;
        this.marketAddress = config.marketAddress;
        if (config.accountSubscription.type === 'polling') {
            this.subscriptionType = 'polling';
            this.accountLoader = config.accountSubscription.accountLoader;
        }
        else {
            this.subscriptionType = 'websocket';
        }
        this.lastSlot = 0;
        this.lastUnixTimestamp = 0;
    }
    async subscribe() {
        if (this.subscribed) {
            return;
        }
        this.market = (0, phoenix_sdk_1.deserializeMarketData)((await this.connection.getAccountInfo(this.marketAddress, 'confirmed'))
            .data);
        const clock = (0, phoenix_sdk_1.deserializeClockData)((await this.connection.getAccountInfo(web3_js_1.SYSVAR_CLOCK_PUBKEY, 'confirmed'))
            .data);
        this.lastUnixTimestamp = (0, phoenix_sdk_1.toNum)(clock.unixTimestamp);
        if (this.subscriptionType === 'websocket') {
            this.marketCallbackId = this.connection.onAccountChange(this.marketAddress, (accountInfo, _ctx) => {
                this.market = (0, phoenix_sdk_1.deserializeMarketData)(accountInfo.data);
            });
            this.clockCallbackId = this.connection.onAccountChange(web3_js_1.SYSVAR_CLOCK_PUBKEY, (accountInfo, ctx) => {
                this.lastSlot = ctx.slot;
                const clock = (0, phoenix_sdk_1.deserializeClockData)(accountInfo.data);
                this.lastUnixTimestamp = (0, phoenix_sdk_1.toNum)(clock.unixTimestamp);
            });
        }
        else {
            this.marketCallbackId = await this.accountLoader.addAccount(this.marketAddress, (buffer, slot) => {
                this.lastSlot = slot;
                this.market = (0, phoenix_sdk_1.deserializeMarketData)(buffer);
            });
            this.clockCallbackId = await this.accountLoader.addAccount(web3_js_1.SYSVAR_CLOCK_PUBKEY, (buffer, slot) => {
                this.lastSlot = slot;
                const clock = (0, phoenix_sdk_1.deserializeClockData)(buffer);
                this.lastUnixTimestamp = (0, phoenix_sdk_1.toNum)(clock.unixTimestamp);
            });
        }
        this.subscribed = true;
    }
    getBestBid() {
        const ladder = (0, phoenix_sdk_1.getMarketUiLadder)(this.market, this.lastSlot, this.lastUnixTimestamp, 1);
        const bestBid = ladder.bids[0];
        if (!bestBid) {
            return undefined;
        }
        return new anchor_1.BN(Math.floor(bestBid[0] * numericConstants_1.PRICE_PRECISION.toNumber()));
    }
    getBestAsk() {
        const ladder = (0, phoenix_sdk_1.getMarketUiLadder)(this.market, this.lastSlot, this.lastUnixTimestamp, 1);
        const bestAsk = ladder.asks[0];
        if (!bestAsk) {
            return undefined;
        }
        return new anchor_1.BN(Math.floor(bestAsk[0] * numericConstants_1.PRICE_PRECISION.toNumber()));
    }
    getL2Bids() {
        return this.getL2Levels('bids');
    }
    getL2Asks() {
        return this.getL2Levels('asks');
    }
    *getL2Levels(side) {
        // @ts-ignore
        const basePrecision = Math.pow(10, this.market.header.baseParams.decimals);
        const pricePrecision = numericConstants_1.PRICE_PRECISION.toNumber();
        const ladder = (0, phoenix_sdk_1.getMarketUiLadder)(this.market, this.lastSlot, this.lastUnixTimestamp, 20);
        for (let i = 0; i < ladder[side].length; i++) {
            const [priceNum, sizeNum] = ladder[side][i];
            const size = new anchor_1.BN(Math.floor(sizeNum * basePrecision));
            yield {
                price: new anchor_1.BN(Math.floor(priceNum * pricePrecision)),
                size,
                sources: {
                    phoenix: size,
                },
            };
        }
    }
    async unsubscribe() {
        if (!this.subscribed) {
            return;
        }
        // remove listeners
        if (this.subscriptionType === 'websocket') {
            await this.connection.removeAccountChangeListener(this.marketCallbackId);
            await this.connection.removeAccountChangeListener(this.clockCallbackId);
        }
        else {
            this.accountLoader.removeAccount(this.marketAddress, this.marketCallbackId);
            this.accountLoader.removeAccount(web3_js_1.SYSVAR_CLOCK_PUBKEY, this.clockCallbackId);
        }
        this.subscribed = false;
    }
}
exports.PhoenixSubscriber = PhoenixSubscriber;
